### ISSUE TITLE
blob/all: ability to use versioned resources

### ISSUE THREAD

**Issue by zachgersh** [google/go-cloud#1343]
### Is your feature request related to a problem? Please describe.

It seems that all of the major cloud providers supported by go-cloud have a concept of versioned download / listing of blobs. I may have missed it but there doesn't seem to be a way to list or download versioned blobs via go-cloud itself.

### Describe the solution you'd like

Let's come up with some primitives that allow users to manipulate the contents of a versioned bucket in some way.

### Describe alternatives you've considered

None. As long as the versioning APIs are somewhat close we can probably design something to work here.

### Additional context

If we can rough out a design here (if we think the feature is valuable) I'd be glad to PR it. Thought of just starting with a PR but didn't want to be presumptuous in case this wasn't a feature that was wanted.

**Comment by vangent** [comment]
Hi @zachgersh , thanks for the suggestion!

Do you mean the ability to list blobs in a bucket, including multiple versions of any given blob? E.g., for AWS, something like:

https://godoc.org/github.com/aws/aws-sdk-go/service/s3#S3.ListObjectVersions

Or, would the ability to request versions for a specific blob be sufficient? What would you do with the non-live blobs? It would help use to understand your use case.

**Comment by zachgersh** [comment]
@vangent - thanks a ton for the quick reply.

Sorry for my lack of info early, wanted to see if this was something that y'all were even interested in first.

Yes, `ListingObjectVersions` is definitely one thing I wanted to implement. 

Along with these others:

Getting the Bucket Versioning:
https://godoc.org/github.com/aws/aws-sdk-go/service/s3#S3.GetBucketVersioning

Ability to Delete an specific version of an Object (so passing VersionID to delete input)
https://github.com/aws/aws-sdk-go/blob/c083ef3c7b58a0b80af22417b42be621f3526cdb/service/s3/api.go#L10446

Download a specific Version using the HeadObject in the request:
https://github.com/aws/aws-sdk-go/blob/c083ef3c7b58a0b80af22417b42be621f3526cdb/service/s3/api.go#L14147

So, basically covering all of the versioning use cases around Creating / Listing / Deleting.

**Comment by vangent** [comment]
@zachgersh Do you mind if I ask what your use case for doing this is?

It seems like a fairly major API change -- we'd either need to amend several existing operations (at least read, delete, list) to support versions, or we'll need to add parallel operations for each of them that are specific to versioned blobs. Do you have an idea for which of those you'd suggest?

We'd also need to confirm that AWS / GCS / Azure Storage all support versioned blobs in a way that can be represented in a unified way. I looked briefly; AWS uses a string, GCS uses an int64 (and has a separate versionID for metadata, as opposed to content), Azure uses a DateTime.

I think the next step is for you to propose what your suggested API changes to the concrete type would be. Depending on how intrusive the changes are, we'll have to decide whether it's worth adding this.

**Comment by zachgersh** [comment]
@vangent - use case for me is that a CI system I use pretty heavily @ Pivotal has a concept of a "resource" which essentially can either fetch or push data. S3 is one of those resources but it has three implementations:

- https://github.com/pivotal-cf/azure-blobstore-resource
- https://github.com/concourse/s3-resource
- https://github.com/frodenas/gcs-resource

I'd like to unify these into one resource using go-cloud but each resource relies on that fact that versioning is supported (since we need a specific version in CI)

My preference would be to amend in a way that allows everyone to get this for free _but_ I could also see a separate set of versioned calls that people can use when interacting with a resource that has a version.

Really appreciate your research. Let me see if I can come up with an API that looks somewhat alright based on what I mentioned above.

**Comment by vangent** [comment]
Great, thanks! Please assign back to me when you have an API proposal. You can do it as a PR, or as a doc, or as a comment here, whichever is easier.

**Comment by zachgersh** [comment]
@vangent will do!

**Comment by eliben** [comment]
Back in 2019 Github still didn't have the feature of distinguishing *why* an issue was closed. Not sure why it back-labels this close as "completed". It's possible this issue was closed due to staleness since the request didn't end up with a proposal after several months. It's OK to reopen issues if the topic arises again and there's fresh user input.
